File: carezoa-backend/scripts/seed.py
"""
Demo seed: 10 nurses, 10 patients, service catalogue, bookings in EVERY state,
a completed visit with report+payout, flagged chat sample, care packages.

Run:  python -m scripts.seed   (from carezoa-backend/, with DATABASE_URL set)
"""

from __future__ import annotations

import asyncio
from datetime import datetime, timedelta

from sqlalchemy import func, select

from app.core.security import Role
from app.db.session import SessionFactory
from app.models.booking import Booking, Review, ServiceReport
from app.models.catalog import Service
from app.models.care_plan import CarePackage
from app.models.communication import CommunicationEvent, CommunicationThread
from app.models.patient import PatientProfile
from app.models.payment import GatewayPaymentStatus, Payment, Payout, PayoutStatus
from app.models.provider import Provider, ProviderAvailability, ProviderServiceOffering, VerificationStatus
from app.models.support import AuditLog
from app.models.user import User
from app.state_machines.booking_state_machine import BookingStatus, PaymentStatus

NOW = datetime.utcnow


def h(hours: float) -> datetime:
    return NOW() + timedelta(hours=hours)


SERVICES = [
    ("Nursing", "Home Nursing Visit", "Registered nurse home visit.", 60, 799, "medkit"),
    ("Elder Care", "Elder Companion Care", "Attendant companionship + routine.", 180, 1199, "heart"),
    ("Recovery", "Post-Surgery Care", "Structured post-op visits.", 90, 1499, "bandage"),
    ("Recovery", "Physiotherapy Session", "Licensed physio at home.", 45, 999, "walk"),
    ("Procedures", "Injection / IV at Home", "Safe injectables at home.", 30, 499, "eyedrop"),
    ("Elder Care", "Attendant — Day Shift", "12-hour attendant shift.", 720, 1899, "people"),
]

NURSES = [
    ("Suresh Patra", "Critical Care Nurse", 9, 4.9, 132, (20.3525, 85.8194), 15, 0.97, 0.01),
    ("Anita Das", "Elder Care Specialist", 7, 4.8, 98, (20.3311, 85.8345), 12, 0.95, 0.02),
    ("Priyanka Sahoo", "Post-Op Recovery Nurse", 6, 4.9, 76, (20.2961, 85.8245), 18, 0.96, 0.01),
    ("Ramesh Behera", "Physiotherapist", 11, 4.7, 210, (20.4625, 85.8830), 25, 0.92, 0.04),
    ("Kavita Mishra", "Procedures Nurse", 5, 4.6, 54, (20.3152, 85.8085), 10, 0.9, 0.05),
    ("Debasish Lenka", "ICU Nurse", 8, 4.8, 140, (20.3402, 85.8097), 20, 0.94, 0.03),
    ("Subhasree Nayak", "Geriatric Nurse", 12, 4.8, 88, (30.3165, 78.0322), 15, 0.93, 0.02),
    ("Bikram Sahu", "Home Care Attendant", 4, 4.5, 41, (20.3481, 85.8420), 12, 0.88, 0.06),
    ("Rita Mohapatra", "Pediatric Home Nurse", 6, 4.7, 63, (20.3205, 85.8170), 14, 0.91, 0.03),
    ("Manas Pradhan", "Rehab Specialist", 10, 4.6, 77, (20.3010, 85.8392), 22, 0.9, 0.04),
]

PATIENTS = [
    "Maya Mohanty", "Rohit Jena", "Sulochana Das", "Biren Mohanty", "Anu Sharma",
    "Prakash Sahoo", "Lina Tripathy", "Daniel D'Souza", "Farah Khan", "Gopal Behera",
]

OFFER_MAP = {
    "Critical Care Nurse": [0, 2, 4], "Elder Care Specialist": [1, 5],
    "Post-Op Recovery Nurse": [0, 2], "Physiotherapist": [3],
    "Procedures Nurse": [0, 4], "ICU Nurse": [0, 2],
    "Geriatric Nurse": [0, 1, 5], "Home Care Attendant": [1, 5],
    "Pediatric Home Nurse": [0], "Rehab Specialist": [2, 3],
}


async def seed() -> None:
    async with SessionFactory() as s:
        if (await s.execute(select(func.count()).select_from(User))).scalar_one() > 0:
            print("already seeded")
            return

        # Create admin user
        admin_user = User(phone="+919999999999", role=Role.ADMIN, mfa_enabled=True)
        s.add(admin_user)
        
        # Create support agent user
        support_user = User(phone="+919999999998", role=Role.SUPPORT_AGENT, mfa_enabled=False)
        s.add(support_user)
        
        await s.flush()
        print(f"Created admin user: {admin_user.phone}")
        print(f"Created support agent: {support_user.phone}")

        services = [
            Service(category=c, name=n, description=d, duration_min=du, base_price_inr=p, icon=i)
            for c, n, d, du, p, i in SERVICES
        ]
        s.add_all(services)
        await s.flush()

        patient_profiles: list[PatientProfile] = []
        for i, name in enumerate(PATIENTS):
            user = User(phone=f"+9194370000{i:02d}", role=Role.PATIENT)
            s.add(user)
            await s.flush()
            profile = PatientProfile(
                user_id=user.id,
                name=name,
                dob="14/09/1971",
                gender="F" if i % 2 == 0 else "M",
                city="Bhubaneswar",
                address=f"{i + 11} Lakeview Lane, Patia",
                lat=20.3525,
                lng=85.8305,
                onboarding_done=True,
            )
            s.add(profile)
            patient_profiles.append(profile)
        await s.flush()

        providers: list[Provider] = []
        for i, (name, title, years, rating, count, (lat, lng), cov, acc, canc) in enumerate(NURSES):
            user = User(phone=f"+9194470000{i:02d}", role=Role.PROVIDER)
            s.add(user)
            await s.flush()
            provider = Provider(
                user_id=user.id,
                display_name=name,
                title=title,
                qualifications=["Registered"],
                languages=["Odia", "Hindi"],
                city="Bhubaneswar",
                lat=lat,
                lng=lng,
                coverage_km=cov,
                years_exp=years,
                rating_avg=rating,
                rating_count=count,
                acceptance_rate=acc,
                cancellation_rate=canc,
                verification_status=VerificationStatus.VERIFIED if i < 8 else VerificationStatus.PENDING_REVIEW,
                bio=f"{title} serving Bhubaneswar for {years} years.",
            )
            s.add(provider)
            providers.append(provider)
        await s.flush()

        for provider in providers:
            for svc_idx in OFFER_MAP.get(provider.title, [0]):
                s.add(
                    ProviderServiceOffering(
                        provider_id=provider.id,
                        service_id=services[svc_idx].id,
                        price_inr=services[svc_idx].base_price_inr,
                    )
                )
            for weekday in range(0, 6):
                s.add(
                    ProviderAvailability(
                        provider_id=provider.id, weekday=weekday, start_min=7 * 60, end_min=20 * 60
                    )
                )
        await s.flush()

        # ---- bookings in every state --------------------------------------
        def mk(idx: int, status: BookingStatus, start: datetime, **kw) -> Booking:
            provider = providers[idx % len(providers)]
            svc = services[idx % len(services)]
            return Booking(
                patient_id=patient_profiles[idx % len(patient_profiles)].id,
                provider_id=provider.id,
                service_id=svc.id,
                status=status,
                payment_status=(
                    PaymentStatus.UNPAID
                    if status == BookingStatus.PENDING_PAYMENT
                    else PaymentStatus.PAID
                ),
                starts_at=start,
                duration_min=svc.duration_min,
                patient_snapshot={"name": PATIENTS[idx % len(PATIENTS)], "age": 68, "gender": "M"},
                address="Flat 4B, Silver Oak Residency, Patia",
                city="Bhubaneswar",
                amount_inr=svc.base_price_inr,
                **kw,
            )

        s.add_all(
            [
                mk(0, BookingStatus.PENDING_PAYMENT, h(30)),                  # awaiting payment
                mk(1, BookingStatus.CONFIRMED, h(20)),                          # confirmed
                mk(2, BookingStatus.EN_ROUTE, h(0.5), en_route_at=NOW(), checkin_otp="4821"),
                mk(3, BookingStatus.CHECKED_IN, NOW(), en_route_at=h(-1), checked_in_at=NOW(), checkin_otp="9310"),
                mk(4, BookingStatus.IN_SERVICE, NOW(), started_at=NOW(), checkin_otp="1042"),
                mk(
                    5, BookingStatus.COMPLETED, h(-26),
                    en_route_at=h(-28), checked_in_at=h(-27.5), started_at=h(-27), completed_at=h(-25.5),
                ),
                mk(6, BookingStatus.CANCELLED, h(-50), cancel_reason="Family rescheduled"),
                mk(7, BookingStatus.NO_SHOW, h(-70), confirmed_at=h(-96)),
                mk(8, BookingStatus.DISPUTED, NOW(), started_at=h(-1)),
            ]
        )
        await s.flush()

        # completed booking → payment + payout (READY after report) + review + report
        completed = (
            await s.execute(select(Booking).where(Booking.status == BookingStatus.COMPLETED).limit(1))
        ).scalar_one()
        payment = Payment(booking_id=completed.id, amount_inr=completed.amount_inr, status=GatewayPaymentStatus.SUCCESS)
        s.add(payment)
        await s.flush()
        s.add(
            ServiceReport(
                booking_id=completed.id,
                summary="Vitals stable, medication schedule reviewed with the family.",
                vitals={"BP": "126/82", "Pulse": "76", "SpO2": "98%"},
                notes="Next visit in 3 days.",
            )
        )
        s.add(
            Payout(
                provider_id=completed.provider_id,
                booking_id=completed.id,
                amount_inr=int(completed.amount_inr * 0.85),
                platform_fee_inr=int(completed.amount_inr * 0.15),
                status=PayoutStatus.READY,
                ready_at=NOW(),
            )
        )
        s.add(
            Review(
                booking_id=completed.id,
                patient_id=completed.patient_id,
                provider_id=completed.provider_id,
                rating=5,
                text="Wonderful nurse, very punctual.",
            )
        )

        # flagged chat sample (anti-bypass review queue)
        thread = CommunicationThread(booking_id=completed.id)
        s.add(thread)
        await s.flush()
        s.add_all(
            [
                CommunicationEvent(
                    thread_id=thread.id, sender_user_id=completed.provider_id or 1,
                    author_name="Nurse", body="I will reach by 9:50 tomorrow.",
                ),
                CommunicationEvent(
                    thread_id=thread.id, sender_user_id=completed.patient_id,
                    author_name="Family", body="call me on [hidden by CAREZOA] directly",
                    flagged=True, flag_patterns=["phone_number"],
                ),
            ]
        )

        s.add_all(
            [
                CarePackage(name="Elder Essentials", description="Companionship + wellness checks.", visits_per_month=8, price_per_month_inr=3999, includes=["4 companion visits", "4 nursing check-ins"], best_for="Independent elders"),
                CarePackage(name="Post-Op Recovery", description="First-month post-surgery pathway.", visits_per_month=10, price_per_month_inr=5499, includes=["6 nursing visits", "2 physio sessions"], best_for="Fresh from discharge"),
                CarePackage(name="Chronic Care Plus", description="Ongoing condition management.", visits_per_month=12, price_per_month_inr=6999, includes=["8 nursing visits", "4 attendant shifts"], best_for="Long-term conditions"),
            ]
        )

        s.add(
            AuditLog(
                actor_user_id=None,
                actor_role="system",
                entity_type="seed",
                entity_id=0,
                action="seed.completed",
                meta={
                    "nurses": len(NURSES),
                    "patients": len(PATIENTS),
                    "admin_phone": "+919999999999",
                    "support_phone": "+919999999998",
                },
            )
        )
        await s.commit()
        print("seeded: 10 nurses, 10 patients, 9 bookings across every state")
        print("admin login: +919999999999 (OTP: 123456 in dev)")
        print("support login: +919999999998 (OTP: 123456 in dev)")


if __name__ == "__main__":
    asyncio.run(seed())

File: src/server/carezoa/seed.ts
import { db } from "@/db";
import {
  czBookings,
  czFamily,
  czMessages,
  czPackages,
  czPatients,
  czPaymentMethods,
  czProviders,
  czRecords,
  czReviews,
  czServices,
  czTickets,
} from "@/db/schema";
import { sql } from "drizzle-orm";

const DAY = 86_400_000;
const HOUR = 3_600_000;

function at(daysFromNow: number, hhmm: string) {
  const [h, m] = hhmm.split(":").map(Number);
  const d = new Date(Date.now() + daysFromNow * DAY);
  d.setHours(h, m, 0, 0);
  return d;
}

let seedPromise: Promise<void> | null = null;
export function ensureCarezoaSeed(): Promise<void> {
  if (!seedPromise) {
    seedPromise = seed().catch((e) => {
      seedPromise = null;
      throw e;
    });
  }
  return seedPromise;
}

async function seed() {
  const existing = await db.execute(sql`select id from cz_patients limit 1`);
  if (existing.rows.length > 0) return;

  await db.insert(czPatients).values({
    id: 1,
    name: "Maya Mohanty",
    phone: "+919437000001",
    dob: "08/03/1992",
    gender: "F",
    city: "Bhubaneswar",
    address: "Flat 4B, Silver Oak Residency, Patia",
    onboardingDone: true,
  });

  await db.insert(czServices).values([
    { category: "Nursing", name: "Home Nursing Visit", description: "Registered nurse home visit: vitals, medications, wound care, catheter care.", durationMin: 60, basePriceInr: 799, icon: "medkit" },
    { category: "Elder Care", name: "Elder Companion Care", description: "Trained attendant for companionship, meals, mobility help and daily routine.", durationMin: 180, basePriceInr: 1199, icon: "heart" },
    { category: "Recovery", name: "Post-Surgery Care", description: "Structured post-op recovery visits: dressing, pain protocol, mobility plan.", durationMin: 90, basePriceInr: 1499, icon: "bandage" },
    { category: "Recovery", name: "Physiotherapy Session", description: "Licensed physiotherapist session at home with exercise plan.", durationMin: 45, basePriceInr: 999, icon: "walk" },
    { category: "Procedures", name: "Injection / IV at Home", description: "Safe administration of prescribed injections or IV drips at home.", durationMin: 30, basePriceInr: 499, icon: "eyedrop" },
    { category: "Elder Care", name: "Attendant — Day Shift", description: "Verified attendant for a full 12-hour day shift with an elderly patient.", durationMin: 720, basePriceInr: 1899, icon: "people" },
  ]);

  const [suresh, anita, priyanka, ramesh, kavita, debasish] = await db
    .insert(czProviders)
    .values([
      { name: "Suresh Patra", title: "Critical Care Nurse", qualifications: ["BSc Nursing", "ICU Certified"], languages: ["Odia", "Hindi", "English"], city: "Bhubaneswar", lat: 20.3525, lng: 85.8194, coverageKm: 15, rating: 4.9, reviewsCount: 132, yearsExp: 9, bio: "Ex-AMRI ICU nurse. Specialises in post-cardiac and ventilator-stepdown home care.", photoColor: "moss", nextAvailableAt: at(0, "18:00") },
      { name: "Anita Das", title: "Elder Care Specialist", qualifications: ["GNM", "Dementia Care Trained"], languages: ["Odia", "Hindi"], city: "Bhubaneswar", lat: 20.3311, lng: 85.8345, coverageKm: 12, rating: 4.8, reviewsCount: 98, yearsExp: 7, bio: "Gentle, patient-first elder companion care. Experienced with Parkinson's and post-stroke routines.", photoColor: "gold" },
      { name: "Priyanka Sahoo", title: "Post-Op Recovery Nurse", qualifications: ["MSc Nursing"], languages: ["Odia", "English"], city: "Bhubaneswar", lat: 20.2961, lng: 85.8245, coverageKm: 18, rating: 4.9, reviewsCount: 76, yearsExp: 6, bio: "Surgical recovery pathway specialist — ortho, abdominal and C-section aftercare.", photoColor: "lilac" },
      { name: "Ramesh Behera", title: "Physiotherapist", qualifications: ["BPT", "MPT (Ortho)"], languages: ["Odia", "Hindi", "English"], city: "Cuttack", lat: 20.4625, lng: 85.883, coverageKm: 25, rating: 4.7, reviewsCount: 210, yearsExp: 11, bio: "Home-based rehab for knee/hip replacement, stroke and sports injuries.", photoColor: "sky" },
      { name: "Kavita Mishra", title: "Procedures Nurse", qualifications: ["ANM", "IV Therapy Certified"], languages: ["Odia"], city: "Bhubaneswar", lat: 20.3152, lng: 85.8085, coverageKm: 10, rating: 4.6, reviewsCount: 54, yearsExp: 5, bio: "Quick, careful injectables and IV administration at home.", photoColor: "blush" },
      { name: "Debasish Lenka", title: "ICU Nurse", qualifications: ["BSc Nursing", "ACLS"], languages: ["Odia", "Hindi", "English"], city: "Bhubaneswar", lat: 20.3402, lng: 85.8097, coverageKm: 20, rating: 4.8, reviewsCount: 140, yearsExp: 8, bio: "High-acuity home support: tracheostomy, NG feeds, advanced wound care.", photoColor: "ink" },
    ])
    .returning({ id: czProviders.id });

  await db.insert(czReviews).values([
    { providerId: suresh.id, authorName: "S. Mohanty", rating: 5, text: "Suresh managed my father's post-bypass care flawlessly. Explained everything to us on video call since we live in Bangalore.", createdAt: at(-22, "11:00") },
    { providerId: suresh.id, authorName: "R. Tripathy", rating: 5, text: "Punctual, calm, extremely competent with IV and dressings.", createdAt: at(-9, "16:30") },
    { providerId: anita.id, authorName: "P. Das", rating: 5, text: "My mother waits for Anita's visits all week. She treats her like family.", createdAt: at(-15, "09:20") },
    { providerId: anita.id, authorName: "K. Jena", rating: 4, text: "Very kind and reliable attendant for evening companionship.", createdAt: at(-4, "18:00") },
    { providerId: priyanka.id, authorName: "M. Rao", rating: 5, text: "Post knee-replacement recovery was smooth under her plan.", createdAt: at(-12, "13:00") },
    { providerId: ramesh.id, authorName: "D. Panda", rating: 4, text: "Strict but effective physio. Dad is walking without support now.", createdAt: at(-7, "10:15") },
  ]);

  const svc = await db.select().from(czServices);
  const byName = Object.fromEntries(svc.map((s) => [s.name, s]));

  const [b1, b2, b3] = await db
    .insert(czBookings)
    .values([
      {
        providerId: anita.id,
        serviceId: byName["Elder Companion Care"].id,
        status: "confirmed",
        paymentStatus: "paid",
        startsAt: at(1, "10:00"),
        durationMin: 180,
        patientName: "Biren Mohanty",
        patientAge: 68,
        patientGender: "M",
        address: "Flat 4B, Silver Oak Residency, Patia",
        city: "Bhubaneswar",
        instructions: "Father is hard of hearing in the left ear — please speak on his right side. Likes tea at 10:30.",
        amountInr: 1199,
        confirmedAt: at(-1, "17:05"),
      },
      {
        providerId: suresh.id,
        serviceId: byName["Home Nursing Visit"].id,
        status: "completed",
        paymentStatus: "paid",
        startsAt: at(-1, "09:00"),
        durationMin: 60,
        patientName: "Sunita Mohanty",
        patientAge: 64,
        patientGender: "F",
        address: "Flat 4B, Silver Oak Residency, Patia",
        city: "Bhubaneswar",
        instructions: "Post-discharge BP and sugar monitoring.",
        amountInr: 799,
        confirmedAt: at(-2, "19:00"),
        enRouteAt: at(-1, "08:21"),
        checkedInAt: at(-1, "08:55"),
        startedAt: at(-1, "09:02"),
        completedAt: at(-1, "10:05"),
        checkinOtp: "7315",
      },
      {
        providerId: kavita.id,
        serviceId: byName["Injection / IV at Home"].id,
        status: "completed",
        paymentStatus: "paid",
        startsAt: at(-4, "18:00"),
        durationMin: 30,
        patientName: "Biren Mohanty",
        patientAge: 68,
        patientGender: "M",
        address: "Flat 4B, Silver Oak Residency, Patia",
        city: "Bhubaneswar",
        instructions: "Prescribed B12 injection — prescription attached in records.",
        amountInr: 499,
        confirmedAt: at(-4, "12:00"),
        enRouteAt: at(-4, "17:32"),
        checkedInAt: at(-4, "17:58"),
        startedAt: at(-4, "18:03"),
        completedAt: at(-4, "18:30"),
        checkinOtp: "2049",
      },
    ])
    .returning({ id: czBookings.id });

  await db.insert(czMessages).values([
    { bookingId: b1.id, sender: "provider", authorName: "Anita Das", body: "Namaste Maya ji! I'll reach by 9:50 tomorrow morning. Should I bring anything specific for your father's breakfast routine?", createdAt: at(-1, "18:02") },
    { bookingId: b1.id, sender: "patient", authorName: "Maya Mohanty", body: "Namaste! Nothing needed, the kitchen is stocked. He prefers oats over rice in the morning these days.", createdAt: at(-1, "18:20") },
    { bookingId: b1.id, sender: "provider", authorName: "Anita Das", body: "Noted — oats it is. I'll also take him for a short corridor walk after breakfast if he feels up to it.", createdAt: at(-1, "18:41") },
  ]);

  await db.insert(czRecords).values([
    {
      bookingId: b2.id,
      summary: "Post-discharge monitoring visit completed. Vitals stable; BP improved versus discharge readings. Continued medication as per discharge summary.",
      vitals: { "Blood pressure": "128/84 mmHg", Pulse: "78 bpm", SpO2: "97%", Temperature: "98.2°F", "Blood sugar (random)": "142 mg/dL" },
      notes: "Advised low-salt diet and evening walk. Next review recommended in 5 days.",
    },
    {
      bookingId: b3.id,
      summary: "Vitamin B12 injection administered intramuscularly without complications. Site observed for 15 minutes post-injection — no reaction.",
      vitals: { Pulse: "74 bpm" },
      notes: "Next dose due in 7 days as per prescription.",
    },
  ]);

  await db.insert(czFamily).values([
    { patientId: 1, name: "Biren Mohanty", relation: "Father", phone: "+919438100201", accessScope: { viewVisits: true, viewRecords: true, chat: true }, inviteStatus: "active" },
    { patientId: 1, name: "Sunita Mohanty", relation: "Mother", phone: "+919438100202", accessScope: { viewVisits: true, viewRecords: false, chat: false }, inviteStatus: "pending" },
  ]);

  await db.insert(czPackages).values([
    { name: "Elder Essentials", description: "Routine companionship and wellness checks for independent elders.", visitsPerMonth: 8, pricePerMonthInr: 3999, includes: ["4 companion visits (3h)", "4 nursing check-ins", "Monthly vitals report to family", "Priority same-day booking"], bestFor: "Independent elders living alone" },
    { name: "Post-Op Recovery", description: "Intense first-month recovery pathway after surgery.", visitsPerMonth: 10, pricePerMonthInr: 5499, includes: ["6 post-op nursing visits", "2 physiotherapy sessions", "2 wound-care visits", "Daily check-in calls"], bestFor: "First 4 weeks after discharge" },
    { name: "Chronic Care Plus", description: "Ongoing support for diabetes, hypertension and cardiac conditions.", visitsPerMonth: 12, pricePerMonthInr: 6999, includes: ["8 nursing visits", "4 attendant day-shifts", "Medication management", "Quarterly doctor coordination"], bestFor: "Long-term condition management" },
  ]);

  await db.insert(czPaymentMethods).values([
    { patientId: 1, type: "upi", label: "UPI", detail: "maya.m@okhdfc" },
    { patientId: 1, type: "card", label: "Visa", detail: "•••• 4242" },
  ]);

  await db.insert(czTickets).values([
    { patientId: 1, subject: "Reschedule day-shift timing", body: "Can the attendant shift on the 12th start at 8am instead of 9am? Father has a hospital appointment that morning.", status: "open" },
  ]);
}
